### PR TITLE
Index Elasticsearch if mappings don't exist

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
[SEAB-4598](https://ucsc-cgl.atlassian.net/browse/SEAB-4598)

<s>This is an alternative solution for the ticket. The other solution is described in PR https://github.com/dockstore/dockstore-deploy/pull/560.</s>

This PR indexes Elasticsearch upon startup if the mappings for the tools and workflows indices don't exist.

This solution takes inspiration from https://www.elastic.co/guide/en/elasticsearch/guide/current/concurrency-solutions.html#global-lock to implement a lock that is used for indexing upon startup to prevent multiple webservices from indexing ES at the same time. I tested this locally by starting two webservices at the same time and only one webservice was able to acquire the lock and index ES successfully.

**Issue**
A link to a github issue or SEAB- ticket (using that as a prefix)

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

**Comment by denis-yuen** [comment]
> This solution takes inspiration from https://www.elastic.co/guide/en/elasticsearch/guide/current/concurrency-solutions.html#global-lock to implement a lock that is used for indexing upon startup to prevent multiple webservices from indexing ES at the same time. I tested this locally by starting two webservices at the same time and only one webservice was able to acquire the lock and index ES successfully.

Nice, you found it!

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java [hunk 1]
```diff
@@ -630,6 +630,7 @@ public static class ElasticSearchConfig {
         private String user;
         private String password;
         private Integer maxConcurrentSessions;
+        private boolean isIndexOnStart = false;
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java:633]: is this needed with the lock approach?

**Review by denis-yuen — APPROVED** [review]

**Comment by denis-yuen** [comment]
> If this is the accepted solution, I will update the dockstore-deploy PR and create a PR in compose_setup to add the `isIndexOnStart` key to the config file.

Looks like yes, but not clear to me why (presumably it wouldn't reindex when restarting in a dev environment if elasticsearch is already indexed)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java (hunk 1 above)
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java:633]: This was implemented per Steve's suggestion https://github.com/dockstore/dockstore-deploy/pull/560#discussion_r947075759. 

Turns out it's also needed for a lot of tests to pass without making changes to the test files. In a lot of tests, the Elasticsearch client is not started, so the attempt to check if the mappings exist throws a NPE. With this flag being set as false by default, the tests won't even execute this Elasticsearch code.

**Comment by kathy-t** [comment]
> Looks like yes, but not clear to me why (presumably it wouldn't reindex when restarting in a dev environment if elasticsearch is already indexed)

An advantage of this approach is that you don't need to provide a token, so it's one less place where the Dockstore token is used. Yeah, it wouldn't re-index if ES is already indexed. This will only index if you deleted your indices or if you're deploying Dockstore with a new Elasticsearch stack.

**Review by coverbeck — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java (hunk 1 above)
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java:633]: Oh, this is a bit inconsistent with our general strategy in ElasticListener (same deal with discourse integration). 
https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java#L110 and https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java#L143
log errors but do not throw exceptions when elastic search is not available

**Comment by denis-yuen** [comment]
> An advantage of this approach is that you don't need to provide a token, so it's one less place where the Dockstore token is used. Yeah, it wouldn't re-index if ES is already indexed. This will only index if you deleted your indices or if you're deploying Dockstore with a new Elasticsearch stack.

Hmm, from above, I think I was more asking, "why do we even need isIndexOnStart", can we get rid of it?

**Comment by svonworl** [comment]
> > An advantage of this approach is that you don't need to provide a token, so it's one less place where the Dockstore token is used. Yeah, it wouldn't re-index if ES is already indexed. This will only index if you deleted your indices or if you're deploying Dockstore with a new Elasticsearch stack.
> 
> Hmm, from above, I think I was more asking, "why do we even need isIndexOnStart", can we get rid of it?

It's probably there because Kathy was trying to accommodate my request.  During development, I didn't want to endure a reindex every time I restarted the webservice.  If the webservice doesn't reindex if the indexes already exist, I can probably retool my scripting/flow pretty easily to account for the lack of the flag.  But, I have the vague feeling that there might be situations (during testing?) when we want to prevent the webservice from doing any indexing until we tell it to...

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceApplication.java [hunk 2]
```diff
@@ -449,6 +456,37 @@ public void run(DockstoreWebserviceConfiguration configuration, Environment envi
...
         CacheConfigManager cacheConfigManager = CacheConfigManager.getInstance();
         cacheConfigManager.initCache();
+
+        if (configuration.getEsConfiguration().isIndexOnStart()) {
+            environment.lifecycle().addLifeCycleListener(new LifeCycle.Listener() {
+                @Override
+                public void lifeCycleStarted(LifeCycle event) {
+                    if (!ElasticSearchHelper.doMappingsExist()) {
+                        // A lock is used to prevent concurrent indexing requests in a deployment where multiple webservices start at the same time
+                        if (ElasticSearchHelper.acquireLock()) {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceApplication.java:466]: Just for the sake of completeness, I should note that this code will allow all N tasks to potentially perform an ES index.   If they all pass `doMappingsExist` and then stall out for the right amounts of time before `acquireLock`, each could theoretically acquire the lock and index, one by one.

This situation will basically never happen in real life, and the penalty is minor (>1 nonoverlapping ES indexes), so no change necessary.

**Review by svonworl — APPROVED** [review]

**Comment by kathy-t** [comment]
Removed the `isIndexOnStart` config key and added some tests. Marking as ready for review and re-requesting reviewers

**Review by svonworl — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]

**Pull request merged**